### DISCUSSION TITLE
modify docker daemon configuration in Rancher Desktop VM

### DISCUSSION THREAD

**Discussion by w7089** [rancher-sandbox/rancher-desktop#1477] [Q&A] [answered]
Is there a way to modify docker daemon configuration in Rancher Desktop VM, e.g. adding insecure registries, etc...?

After such modification, I'd like to restart docker daemon for changes to apply. Is there a way to connect to Rancher Desktop VM?

Haven't found anything relevant in https://docs.rancherdesktop.io/faq

**Comment by jandubois** [comment, marked as answer]
You didn't specify the OS, so I'm going to assume macOS because I already have the snippets for it in my clipboard history. 😄 

There is no supported way, but you can ssh into the VM and edit the config file manually to set `DOCKER_OPTS`.

Example from an already configured VM:

```console
$ LIMA_HOME="$HOME/Library/Application Support/rancher-desktop/lima" "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/lima/bin/limactl" shell 0
lima-rancher-desktop:~# tail -1 /etc/conf.d/docker
DOCKER_OPTS="--insecure-registry=insecure.home:80"
```

You could restart the docker service afterwards, but stoping and restarting Rancher Desktop would be the clean way to make sure everything including kubernetes is restarted correctly.

**Reply by electriquo** [reply]
> you can ssh into the VM

Could you share how?

**Reply by jandubois** [reply]
It is literally the first line in the snippet above:

```console
LIMA_HOME="$HOME/Library/Application Support/rancher-desktop/lima" "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/lima/bin/limactl" shell 0
```

😄

**Reply by Alexhha** [reply]
@jandubois 

> You could restart the docker service afterwards, but stoping and restarting Rancher Desktop would be the clean way to make sure everything including kubernetes is restarted correctly.

were you able to run `docker login` after that ? If so, to what address did you point `insecure.home` name ?

**Reply by bneradt** [reply]
This used to work for me, but both the `/etc/conf.d/docker` and the `/etc/docker/daemon.json` get rewritten to their original values after I restart Rancher through the UI. Is there a way to get these files to persist now?

**Comment by pavel-kalmykov** [comment]
In case someone finds this, another possibility instead of modifying `/etc/conf.d/docker` is to create a `/etc/docker/daemon.json` file with your desired configuration. Once you enter the Lima VM as @jandubois suggested:

```console
$ LIMA_HOME="$HOME/Library/Application Support/rancher-desktop/lima" "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/lima/bin/limactl" shell 0
```

> PS @jandubois is it possible for Rancher Desktop to provide an easier method to access this VM?

Then, create your `/etc/docker/daemon.json` with the content you would like to have:

```json
{
  "bip": "192.168.98.5/24",
  "insecure-registries" : ["insecure.registry"]
}
```

Don't forget to restart Rancher Desktop to apply these changes!

**Reply by jandubois** [reply]
> > PS @jandubois is it possible for Rancher Desktop to provide an easier method to access this VM?

It is not really an officially supported way to work with Rancher Desktop, so not sure if we need/want to make it easier. You can define a shell alias or wrapper script easily yourself (e.g. I have an `rdlima` alias for it), so I'm also not sure what else we could/should do.

There are so many things that RD does not yet expose through the UI, so shell access is often required. But it comes with the usual caveat: If you break anything, you get to keep all the pieces!

**Reply by giovanni-nabil** [reply]
this file gets re-written if I restart Rancher desktop, so instead i just restarted docker inside lima VM using `sudo service docker restard`

**Comment by wstr-ncrum** [comment]
EDIT

I mistakenly created daemon.js instead of daemon.json. Below is how to access the WSL instance on Windows to add an insecure registry.

```powershell
# You can start an interactive shell
wsl -d rancher-desktop -e /bin/sh

# Or just create the file right away
wsl -d rancher-desktop -e vi /etc/docker/daemon.json
```

Afterwards restart rancher desktop and it should work. If you still have trouble you can try repeating the steps above to create a file at `/etc/default/docker` with the contents:

```sh
DOCKER_OPTS="--config-file=/etc/docker/daemon.json"
```

Then restarted rancher desktop again

Using:
Windows 10 Pro
Rancher-Desktop 1.0.1
Moby(dockerd) runtime

**Comment by bakejeyner** [comment]
I am attempting something similar by trying to insecurely expose the docker daemon on tcp://0.0.0.0:2375. (I'm transitioning away from Docker Desktop and some of our build tools are attempting to communicate with the docker deamon over tcp.) I'm using Rancher Desktop (1.0.1) on Windows 10 (10.0.19044 Build 19044) with an Ubuntu (20.04) WSL2 integration.

[I tried following the more windows-specific advice given by wstr-ncrum](https://github.com/rancher-sandbox/rancher-desktop/discussions/1477#discussioncomment-2128647). I started the interactive shell to the VM and added an `/etc/docker/daemon.json` file containing:
```
{
  "hosts": [
    "tcp://0.0.0.0:2375"
  ]
}
```
[which should be valid according to Docker's documentation](https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-configuration-file). However my machine is still not listening on port 2375 even after adding `/etc/default/docker` and `/etc/conf.d/docker` files to the VM, both containing:
```
DOCKER_OPTS="--config-file=/etc/docker/daemon.json"
```

I'm getting lost in the sauce and could really use some help.

P.S. I think making the docker daemon more easily configurable would really help Rancher Desktop stand out in the great Docker Desktop migration.

**Reply by wstr-ncrum** [reply]
Did you have to insecurely expose the docker daemon before with Docker Desktop for your specific build tools?

**Reply by wstr-ncrum** [reply]
Also did you quit and restart Rancher Desktop after the config changes?

**Reply by bakejeyner** [reply]
>Did you have to insecurely expose the docker daemon before with Docker Desktop for your specific build tools?

I don't know if we _had to_, but doing so was the easiest route. On a related note: in my current effort when I tried using the [maven-dockerfile plugin](https://github.com/spotify/dockerfile-maven) (the build tool in question) with my `DOCKER_HOST=npipe:////./pipe/docker_engine` environment variable set, the plugin complained that there was no hostname.

>Also did you quit and restart Rancher Desktop after the config changes?

Oh ya; I'm a firm believer in turning things off and back on 😄

**Reply by jialecao** [reply]
Hi @wstr-ncrum and @bakejeyner 
I got the following error after I added /etc/daemon/daemon.json and restarted Rancher Desktop. My Rancher Desktop version is 1.4.1. Do I have to do something else? Thank you in advance!
`unable to configure the Docker daemon with file /etc/docker/daemon.json: the following directives are specified both as a flag and in the configuration file: hosts: (from flag: [unix:///mnt/wsl/rancher-desktop/run/docker.sock unix:///var/run/docker.sock], from file: [tcp://0.0.0.0:2375])
could not connect to docker: dial unix /mnt/wsl/rancher-desktop/run/docker.sock: connect: connection refused`

**Comment by Mathanraj-Sharma** [comment]
I am using Mac M1 & rancher desktop `1.4.1` for me it worked when I did below

```
LIMA_HOME="$HOME/Library/Application Support/rancher-desktop/lima" "/Applications/Rancher Desktop.app/Contents/Resources/resources/darwin/lima/bin/limactl" shell 0
```

and added `/etc/docker/daemon.json` with the below settings

```
{
  "builder": {
    "gc": {
      "defaultKeepStorage": "20GB",
      "enabled": true
    }
  },
  "features": {
    "buildkit": true
  },
  "experimental": false,
  "dns": ["your custom dns", "8.8.8.8"],
  "bip": "192.168.254.1/24",
  "insecure-registries" : ["URL for your registry"]
}
```

then restart docker daemon inside lima VM using `sudo service docker restart` or restart rancher desktop

**Comment by stampycode** [comment]
none of the above has worked for me on Windows WSL2.

- on restarting rancher-desktop, the file at `/etc/conf.d/docker` is restored to original state, my changes wiped.
- on restarting docker within `rdctl shell`, the flags in `/etc/conf.d/docker` don't seem to have any effect.

`/etc/docker/daemon.json`
```json
{
  "features": {
    "containerd-snapshotter": true
  },
  "experimental": true,
  "max-concurrent-downloads": 1,
  "max-concurrent-uploads": 1,
  "max-download-attempts": 1
}
```

`/etc/conf.d/docker`
```ini
WSL_HELPER_BINARY="/mnt/c/Program Files/Rancher Desktop/resources/resources/linux/wsl-helper"
DOCKER_OPTS="--config-file=/etc/docker/daemon.json"
LOG_DIR="/mnt/c/Users/me/AppData/Local/rancher-desktop/logs"
```

**Reply by epjane** [reply]
For WSL2 I found the answer [here](https://dev.to/crumdev/replace-docker-with-rancher-desktop-and-wsl2-on-windows-5d16), run this from powershell:
```
wsl -d rancher-desktop -e vi /etc/docker/daemon.json
```
